File: js/queue/arrayQueue.js
const Arr = require("../myArray/myArray");

/**
 * 数组队列的实现
 * 队列的几个核心接口
 * enqueue(elem): 入队一个元素
 * dequeue() elem: 出队一个元素
 * getFront() elem: 查看队首元素
 * done
 */
class ArrayQueue {
  constructor(cap) {
    this.data = new Arr(cap);
  }

  getLength() {
    return this.data.getLength();
  }

  isEmpty() {
    return this.data.isEmpty();
  }

  enqueue(elem) {
    this.data.addLast(elem);
  }

  dequeue() {
    return this.data.removeFirst();
  }

  getFront() {
    return this.data.get(0);
  }

  toString() {
    let str = "Queue front: ";
    str += this.data.toString();
    str += " tail";
    return str;
  }
}

module.exports = ArrayQueue;

File: go/queue/array_queue.go
package queue

import (
	"fmt"
	"../myArray"
)

type ArrayQueue struct {
	data *myArray.Array
}

func NewArrayQueue(arg ...interface{}) (*ArrayQueue, error) {
	if len(arg) == 0 {
		arg = append(arg, 10)
	}
	data, err := myArray.NewArray(arg[0])
	return &(ArrayQueue{data}), err
}

func (q *ArrayQueue) GetSize() int {
	return q.data.GetLength()
}

func (q *ArrayQueue) IsEmpty() bool {
	return q.data.IsEmpty()
}

func (q *ArrayQueue) Enqueue(elem interface{}) {
	q.data.AddLast(elem)
}

// 注意这里有一点偷懒的地方就是 error 并没有做任何的修改，只是简单的把数组的错误再次返回了
func (q *ArrayQueue) Dequeue() (interface{}, error) {
	return q.data.RemoveFirst()
}

func (q *ArrayQueue) GetFront() (interface{}, error) {
	return q.data.Get(0)
}

func (q *ArrayQueue) String() string {
	str := "Queue: Front [ "
	size := q.GetSize()
	for i := 0; i < size; i++ {
		elem, _ := q.data.Get(i)
		str += fmt.Sprint(elem)
		if i != size - 1 {
			str += ", "
		}
	}
	str += " ] Tail"
	return str
}

